### DISCUSSION TITLE
Is there a predefined or preferred way to update the Progress Dialog during CA's or BeforeInstall/AfterInstall code?

### DISCUSSION THREAD

**Discussion by kplatter** [oleg-shilo/wixsharp#1474] [General]
I have code that does different tasks in CA and AfterInstall and would like to update the Progress Dialog so that it indicates what is being done instead of just "Status:         ".  Is there a predefined or preferred way to go about doing this?

I found the following code and it seems to sort of work but does not seem to be completely accurate or complete...

``` C#
            //set the UI Status ACTIONSTART
            Record statusRecordObj = new Record(3);
            statusRecordObj.SetString(1, actionName);
            statusRecordObj.SetString(2, actionDescription);
            //below line defines a template for display on each message tick
            statusRecordObj.SetString(3, "[1]");
            session.Message(WixToolset.Dtf.WindowsInstaller.InstallMessage.ActionStart, statusRecordObj);

            //resets the progress bar.
            Record progressRecordObj = new Record(4);
            progressRecordObj.SetInteger(1, 0); //0 = reset bar
            progressRecordObj.SetInteger(2, numTicksExpected + 1);
            progressRecordObj.SetInteger(3, 0); //0 = left-to-right, 1= R-to-L
            progressRecordObj.SetInteger(4, 0); //0 = calc time remaining
            session.Message(WixToolset.Dtf.WindowsInstaller.InstallMessage.Progress, progressRecordObj);

            //tell the progress bar to increment for each ActionData message
            progressRecordObj.SetInteger(1, 2); //Action Info
            progressRecordObj.SetInteger(2, 1); //numTicks each ActionData
            progressRecordObj.SetInteger(3, 1); //increment bar for each ActionData
            session.Message(WixToolset.Dtf.WindowsInstaller.InstallMessage.Progress, progressRecordObj);
```

**Comment by oleg-shilo** [comment]
I do not see any other way. MSI/WiX does not offer anything to directly support this scenario.

You can have a slight (arguably simpler) modification of your sample though. 
In the progress dialog you are subscribed for all messages but process only progress messages. You can monitor the message flow for a custom message (e.g. a Log message with a special flag in the content) sent from your CA/event. And when received you can update your dialog as you need.

You will need to add a case for `case InstallMessage.Info:`, which I think is a message that is sent when you do `Session.Log(...)`. 

![image](https://github.com/oleg-shilo/wixsharp/assets/16729806/44cf5e0c-b5d0-4f0c-af5b-a01c5123bd4b)
